### ISSUE TITLE
Irregularly spaced custom cycles

### ISSUE THREAD

**Issue by bradyrx** [lukelbd/proplot#84]
I think it would be nice to have an easier way to create irregularly spaced custom cycles. For instance I want 3 colors from 'reds', 4 from 'blues', and so on. Right now if I do

`plot.Cycle('reds', 'blues', 'greens', 15)` it'll put 5 on each without any easy way to alter that.

Here's my hacky way of constructing it:

```python
cycle1 = plot.Cycle('Boreal', 3, left=0.4, right=0.6)
cycle2 = plot.Cycle('Grays', 4, left=0.4, right=0.6)
cycle3 = plot.Cycle('Fire', 5, left=0.4, right=0.6)
cycle4 = plot.Cycle('Glacial', 2, left=0.4, right=0.6)
cycle5 = plot.Cycle('yellow', 1, left=0.4, right=0.6)

cycle = cycle1.concat(cycle2)
cycle = cycle.concat(cycle3)
cycle = cycle.concat(cycle4)
cycle = cycle.concat(cycle5)
```

Perhaps you could do `plot.Cycle('reds', 'blues', 'greens', [3, 4, 5])` or something similar.

**Comment by lukelbd** [comment]
Glad to see someone is using the color features so rigorously! This is actually [sort of already a feature](https://proplot.readthedocs.io/en/latest/api/proplot.styletools.Cycle.html) (see the `samples` argument). A similar example looks something like:

```python
cycle = plot.Cycle('Blues', 'Reds', 'Grays', 'Greens', 'Oranges',
   samples=np.concatenate((
      np.linspace(0.4, 0.6, 3),
      np.linspace(1.4, 1.6, 4),
      np.linspace(2.4, 2.6, 5),
      np.linspace(3.4, 3.6, 2),
      [4.5])) / 5)
```

There are some improvements that could be made though:

* `Cycle` interprets positional arguments with `Colormap`. If Colormap returns a `LinearSegmentedColormap`, it converts it to `ListedColormap` by drawing discrete colors, *registers* that `ListedColormap` under a new name, and then finally creates a `cycler.Cycler` object from the colors in the `ListedColormap`.
  * It might make more sense if I add the option to convert `LinearSegmentedColormap`s to `ListedColormaps` to *Colormap* rather than Cycle, then users can optionally pass *arrays of samples* (e.g. `samples=(3, 4, 5, 2, 1)`) to `Colormap`.
  * Related to this, I couldn't exactly reproduce your example because ProPlot cannot merge `LinearSegmentedColormap`s with `PerceptuallyUniformColormap`s. But I can *also* add an option to Colormap to convert all positional arguments to `LinearSegmentedColormap` before merging.
* Previously I allowed you to apply the `left`, `right`, `cut`, and `reverse` keyword arguments to *each colormap differently* by passing *arrays* instead of scalars. I got rid of this for the sake of simplicity, but maybe should add it back to be consistent with above.

**Comment by lukelbd** [comment]
Don't think I can address this until sometime January but feel free to submit a PR if you want. Now that I have pre-commit and CI set up it should be smoother.

Btw I tried out `black` on a few more files and it mostly won me over (especially with the `-S` flag; like you I prefer single quotes to double quotes). Only annoyance is it sometimes converts long lists like `a = ['a', 'b', 'c', 'd', 'e', 'f']` and function arguments to *one element* per line -- is there a way to change this? I mean I get the motiviation but it often looks kinda ridiculous, especially for commands like `Axes.format` that accept arguments that *naturally* can be grouped with multiple items per line (e.g. `xformatter=None, yformatter=None, xformatter_kw=None, yformatter_kw=None` on one line).

**Comment by bradyrx** [comment]
I'll see if I have time to open a PR, but most likely not. Maybe I can give it a go on the flight home for the holidays. So this should be improved, and also added to the docs once it's improved. 

> Btw I tried out black on a few more files and it mostly won me over (especially with the -S flag; like you I prefer single quotes to double quotes). Only annoyance is it sometimes converts long lists like a = ['a', 'b', 'c', 'd', 'e', 'f'] and function arguments to one element per line -- is there a way to change this? I mean I get the motiviation but it often looks kinda ridiculous, especially for commands like Axes.format that accept arguments that naturally can be grouped with multiple items per line (e.g. xformatter=None, yformatter=None, xformatter_kw=None, yformatter_kw=None on one line).

Whatever they have listed in their options table is the options you can declare. The single line thing is forced when you go over your line limit (we use 88... or 89... I can't remember). I think it's a throwback to Java maybe? I remember this being a style thing in an early language I learned years ago. I think it looks nice, since you can very very clearly see all the options, and then they align with your docstrings. But I could see why it would be annoying when you have a *ton* of keywords, and when they can be grouped together.

They seem open to conversation on the `black` issue threads about changes moving forward. I recall having issues with dictionaries with long keys and had a thread open on that.

**Comment by lukelbd** [comment]
Don't seem to be any style-configuring options other than `--skip-string-normalization` and `--line-length`. Just as an example, black converts

```python
    def format(
        self,
        *,
        aspect=None,
        xloc=None, yloc=None,
        xspineloc=None, yspineloc=None,
        xtickloc=None, ytickloc=None,
        fixticks=False,
        xlabelloc=None, ylabelloc=None,
        xticklabelloc=None, yticklabelloc=None,
        xtickdir=None, ytickdir=None,
        xgrid=None, ygrid=None,
        xgridminor=None, ygridminor=None,
        xtickminor=None, ytickminor=None,
        xticklabeldir=None, yticklabeldir=None,
        xtickrange=None, ytickrange=None,
        xreverse=None, yreverse=None,
        xlabel=None, ylabel=None,
        xlim=None, ylim=None,
        xscale=None, yscale=None,
        xrotation=None, yrotation=None,
        xformatter=None, yformatter=None,
        xticklabels=None, yticklabels=None,
        xticks=None, xminorticks=None,
        xlocator=None, xminorlocator=None,
        yticks=None, yminorticks=None,
        ylocator=None, yminorlocator=None,
        xbounds=None, ybounds=None,
        xmargin=None, ymargin=None,
        xcolor=None, ycolor=None,
        xticklen=None, yticklen=None,
        xlabel_kw=None, ylabel_kw=None,
        xscale_kw=None, yscale_kw=None,
        xlocator_kw=None, ylocator_kw=None,
        xformatter_kw=None, yformatter_kw=None,
        xminorlocator_kw=None, yminorlocator_kw=None,
        patch_kw=None,
        **kwargs,
    ):
```

...to

```python
    def format(
        self,
        *,
        aspect=None,
        xloc=None,
        yloc=None,
        xspineloc=None,
        yspineloc=None,
        xtickloc=None,
        ytickloc=None,
        fixticks=False,
        xlabelloc=None,
        ylabelloc=None,
        xticklabelloc=None,
        yticklabelloc=None,
        xtickdir=None,
        ytickdir=None,
        xgrid=None,
        ygrid=None,
        xgridminor=None,
        ygridminor=None,
        xtickminor=None,
        ytickminor=None,
        xticklabeldir=None,
        yticklabeldir=None,
        xtickrange=None,
        ytickrange=None,
        xreverse=None,
        yreverse=None,
        xlabel=None,
        ylabel=None,
        xlim=None,
        ylim=None,
        xscale=None,
        yscale=None,
        xrotation=None,
        yrotation=None,
        xformatter=None,
        yformatter=None,
        xticklabels=None,
        yticklabels=None,
        xticks=None,
        xminorticks=None,
        xlocator=None,
        xminorlocator=None,
        yticks=None,
        yminorticks=None,
        ylocator=None,
        yminorlocator=None,
        xbounds=None,
        ybounds=None,
        xmargin=None,
        ymargin=None,
        xcolor=None,
        ycolor=None,
        xticklen=None,
        yticklen=None,
        xlabel_kw=None,
        ylabel_kw=None,
        xscale_kw=None,
        yscale_kw=None,
        xlocator_kw=None,
        ylocator_kw=None,
        xformatter_kw=None,
        yformatter_kw=None,
        xminorlocator_kw=None,
        yminorlocator_kw=None,
        patch_kw=None,
        **kwargs,
    ):
```

Will probably use it anyway; *overall* I definitely like what it does to my code. Thanks for telling me about it.

**Comment by lukelbd** [comment]
This is fixed in 972956b1 -- converting `LinearSegmentedColormap` to `ListedColormap` is now handled inside `Colormap`, and you can apply per-colormap arguments for colormap-modifying keywords like `left`, `right`, and `samples`. Here's a quick example:

```python
import proplot as plot
import numpy as np
import pandas as pd
cycle = plot.Cycle(
    'Boreal', 'Grays', 'Fire', 'Glacial', 'yellow',
    left=[0.4] * 5, right=[0.6] * 5,
    samples=[3, 4, 5, 2, 1],
)
fig, ax = plot.subplots()
data = np.random.rand(10, len(cycle)).cumsum(axis=1)
data = pd.DataFrame(data, columns=list('abcdefghijklmno'))
ax.plot(data, cycle=cycle, linewidth=2, legend='b')
```

![tmp](https://user-images.githubusercontent.com/19657652/81490280-b674f500-923d-11ea-9599-0b571d032f90.png)
